### PR TITLE
[EPROD-552] Migrate from ethers-core to alloy

### PR DESCRIPTION
**Pull request by ufoscout**
# Issue ticket

Issue ticket link:
https://infinityswap.atlassian.net/browse/EPROD-552

### Code conventions

- [x] I have performed a self-review of my code
- [x] Every new function is documented
- [x] Object names are auto explicative

### Security

- [ ] The PR does not break APIs backward compatibility
- [x] The PR does not break the stable storage backward compatibility

### Testing

- [x] Every function is properly unit tested
- [x] I have added integration tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] IC endpoints are always tested through the `canister_call!` macro

### CODE REVIEW

File: src/did/src/transaction.rs [hunk 1]
```diff
@@ -1017,10 +1251,10 @@ mod test {
         rand::thread_rng().fill(&mut data);
         let bloom = Bloom(data.into());
 
-        let mut stream = rlp::RlpStream::new();
-        bloom.rlp_append(&mut stream);
-        let encoded = stream.out();
-        let decoded = rlp::decode::<Bloom>(&encoded).unwrap();
+        // let mut stream = rlp::RlpStream::new();
+        // bloom.rlp_append(&mut stream);
```
Comment (Maximkaaa) [inline src/did/src/transaction.rs:1255]: del

File: src/eth-signer/src/sign_strategy.rs [hunk 2]
```diff
@@ -95,55 +78,54 @@ impl Storable for SigningStrategy {
...
-        codec::bincode_decode(&bytes)
-    }
+//     fn from_bytes(bytes: Cow<[u8]>) -> Self {
+//         codec::bincode_decode(&bytes)
+//     }
 
-    const BOUND: Bound = Bound::Unbounded;
-}
+//     const BOUND: Bound = Bound::Unbounded;
+// }
```
Comment (Maximkaaa) [inline src/eth-signer/src/sign_strategy.rs:98]: del

File: src/eth-signer/src/sign_strategy.rs [hunk 3]
```diff
@@ -155,94 +137,93 @@ impl TransactionSigner for TxSigner {
...
+//         let signing_key = SigningKey::from_slice(val.signing_key_bytes)
+//             .map_err(|e| D::Error::custom(format!("failed to decode signing key: {e}")))?;
+//         let address = H160::from_slice(val.address_bytes);
+//         Ok(LocalTxSigner::new(Wallet::new_with_signer(
+//             Cow::Owned(signing_key),
+//             address.into(),
+//             val.chain_id,
+//         )))
+//     }
+// }
```
Comment (Maximkaaa) [inline src/eth-signer/src/sign_strategy.rs:226]: del

File: src/eth-signer/src/sign_strategy.rs [hunk 4]
```diff
@@ -369,54 +349,54 @@ mod test {
...
+    //     }) = signer
+    //     {
+    //         assert!(matches!(key_id, crate::ic_sign::SigningKeyId::Dfx));
+    //         assert_eq!(cached_address, management_canister_signer.cached_address);
+    //         assert_eq!(derivation_path, management_canister_signer.derivation_path);
+    //         assert_eq!(cached_pubkey, management_canister_signer.cached_pubkey)
+    //     } else {
+    //         panic!("roundtrip failed");
+    //     }
+    // }
```
Comment (Maximkaaa) [inline src/eth-signer/src/sign_strategy.rs:399]: del

File: src/eth-signer/src/transaction.rs [hunk 5]
```diff
@@ -1,96 +1,98 @@
...
+    Signature(DidSignature),
+    /// Use signing key to generate signature in `calculate_hash_and_build` method.
+    /// Whenever possible use `LocalWallet` instead of `SigningKey` as it requires less allocations.
     SigningKey(&'a SigningKey),
+    /// Use the wallet to generate signature in `calculate_hash_and_build` method.
+    LocalWallet(&'a LocalWallet),
 }
 
-#[derive(Debug, Clone, PartialEq, Eq)]
+#[derive(Clone)]
```
Comment (Maximkaaa) [inline src/eth-signer/src/transaction.rs:28]: Why would you delete derived traits?

**Review by Maximkaaa — CHANGES_REQUESTED** [review]
Just some mechanical review. I didn't actually check yet what those changes would imply for EVM or Bridges....

**Review by F3kilo — APPROVED** [review]

File: src/did/src/block.rs [hunk 6]
```diff
@@ -157,6 +158,134 @@ impl Block<H256> {
...
+        self.gas_used.encode(out);
+        self.timestamp.encode(out);
+        self.extra_data.encode(out);
+        self.mix_hash.encode(out);
+        self.nonce.encode(out);
+
+        // Encode all the fork specific fields
+        if let Some(ref base_fee) = self.base_fee_per_gas {
+            base_fee.encode(out);
+        }
```
[the remark this answers is not in the document]
Comment (ufoscout) [reply src/did/src/block.rs:195]: We could by creating a temporary Block wrapper that implements the RlpEncodable on itself. I am not sure it worth doing it in this case as the implemtation is pretty simple and I am not sure how the optional `base_fee` field would be managed

File: src/did/src/block.rs [hunk 7]
```diff
@@ -157,6 +158,134 @@ impl Block<H256> {
...
+
+        // No need to encode requests_hash if it doesn't exist.
+        // if let Some(ref requests_hash) = self.requests_hash {
+        //     requests_hash.encode(out);
+        // }
+
+        // No need to encode target_blobs_per_block if it doesn't exist.
+        // if let Some(ref target_blobs_per_block) = self.target_blobs_per_block {
+        //     target_blobs_per_block.encode(out);
+        // }
```
[the remark this answers is not in the document]
Comment (ufoscout) [reply src/did/src/block.rs:225]: done

File: src/did/src/integer.rs [hunk 8]
```diff
@@ -5,61 +5,66 @@ use std::rc::Rc;
...
+    }
+
+    pub fn from_hex_str(s: &str) -> Result<Self, String> {
+        let temp: String;
+        let s = if s.starts_with("0x") || s.starts_with("0X") {
+            s
+        } else {
+            temp = format!("0x{}", s);
+            &temp
+        };
```
[the remark this answers is not in the document]
Comment (ufoscout) [reply src/did/src/integer.rs:66]: The suggested change always allocate a new String, on the contrary, the current implementation allocates only in the worst case.

File: src/did/src/block.rs [hunk 9]
```diff
@@ -196,106 +325,11 @@ impl From<Block<Transaction>> for Block<H256> {
...
-}
-
 /// Calculate the hash of a block
 pub fn calculate_block_hash<T>(block: &Block<T>) -> H256 {
-    let mut rlp = RlpStream::new();
-    block_header_rlp(block, &mut rlp);
-    keccak_hash(&rlp.out())
+    keccak256(block.header_rlp_encoded()).into()
+    // let header: alloy::consensus::Header = block.into();
+    // header.hash_slow().into()
```
[the remark this answers is not in the document]
Comment (ufoscout) [reply src/did/src/block.rs:332]: done

File: src/did/src/integer.rs [hunk 10]
```diff
@@ -120,19 +117,23 @@ impl U256 {
...
```
[the remark this answers is not in the document]
Comment (ufoscout) [reply src/did/src/integer.rs:135]: The suggested change always allocates a new String, on the contrary, the current implementation allocates only in the worst case.

File: src/did/src/integer.rs [hunk 11]
```diff
@@ -541,30 +513,33 @@ mod tests {
...
-        assert_eq!(U256::from(0u64), U256::from_hex_str("00").unwrap());
-        assert_eq!(U256::from(1u64), U256::from_hex_str("01").unwrap());
+        assert_eq!(U256::from(0u64), U256::from_hex_str("0x00").unwrap());
+        assert_eq!(U256::from(1u64), U256::from_hex_str("0X01").unwrap());
+        assert_eq!(U256::from(255u64), U256::from_hex_str("0xff").unwrap());
         assert_eq!(U256::from(255u64), U256::from_hex_str("ff").unwrap());
         assert_eq!(
             U256::from(2074343815918867987178857765017879333u128),
-            U256::from_hex_str("18F810BD8895AA66364CBDD91A20325").unwrap()
+            U256::from_hex_str("0x18F810BD8895AA66364CBDD91A20325").unwrap()
```
[the remark this answers is not in the document]
Comment (ufoscout) [reply src/did/src/integer.rs:530]: I added a specific test for those cases

File: src/did/src/transaction.rs (hunk 1 above)
Comment (ufoscout) [reply src/did/src/transaction.rs:1225]: done

File: src/eth-signer/src/sign_strategy.rs (hunk 3 above)
Comment (ufoscout) [reply src/eth-signer/src/sign_strategy.rs:226]: done

File: src/eth-signer/src/sign_strategy.rs (hunk 4 above)
Comment (ufoscout) [reply src/eth-signer/src/sign_strategy.rs:399]: done

File: src/eth-signer/src/transaction.rs (hunk 5 above)
Comment (ufoscout) [reply src/eth-signer/src/transaction.rs:28]: They're wrong, there's no need for them

**Review by veeso — APPROVED** [review]

**Review by Maximkaaa — APPROVED** [review]

**Pull request merged**